File: sharedComposables/src/commonMain/kotlin/dev/mateusz1913/f1results/composable/common/DriverResultRow.kt
package dev.mateusz1913.f1results.composable.common

import androidx.compose.foundation.border
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.Icon
import androidx.compose.material.MaterialTheme
import androidx.compose.material.Text
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.LocalFireDepartment
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import dev.mateusz1913.f1results.datasource.data.race_results.ResultType

@Composable
fun DriverResultRow(
    result: ResultType,
    noDriverInfo: Boolean = false
) {
    Row(
        modifier = Modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.SpaceBetween,
        verticalAlignment = Alignment.CenterVertically
    ) {
        Row(
            modifier = Modifier.weight(1f),
            horizontalArrangement = Arrangement.SpaceBetween,
            verticalAlignment = Alignment.CenterVertically
        ) {
            StandingPosition(result.positionText)
            Column(modifier = Modifier.weight(1f)) {
                Row(
                    modifier = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.SpaceBetween,
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    if (!noDriverInfo) {
                        Text(
                            "${result.driver.givenName} ${result.driver.familyName}",
                            fontSize = 18.sp,
                            fontWeight = FontWeight.Bold,
                            modifier = Modifier.padding(horizontal = 10.dp, vertical = 2.dp)
                        )
                        Box(modifier = Modifier.padding(horizontal = 10.dp, vertical = 2.dp)) {
                            Column(
                                modifier = Modifier
                                    .border(
                                        1.dp,
                                        MaterialTheme.colors.onBackground,
                                        RoundedCornerShape(13.dp)
                                    )
                                    .size(26.dp),
                                horizontalAlignment = Alignment.CenterHorizontally,
                                verticalArrangement = Arrangement.Center
                            ) {
                                val number = result.number
                                if (number != null) {
                                    Text(number, fontSize = 14.sp, fontWeight = FontWeight.Medium)
                                }
                            }
                        }
                    }
                }
                Row(
                    modifier = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.SpaceBetween,
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    Text(
                        result.constructor.name,
                        fontSize = 16.sp,
                        fontWeight = FontWeight.SemiBold,
                        modifier = Modifier.padding(horizontal = 10.dp, vertical = 2.dp)
                    )
                    Box(modifier = Modifier.padding(vertical = 2.dp)) {
                        Text(
                            result.time?.time ?: result.status,
                            fontSize = 14.sp,
                            fontWeight = FontWeight.Medium,
                            modifier = Modifier.padding(end = 10.dp),
                            textAlign = TextAlign.Right
                        )
                    }
                }
            }
        }
        Row(
            horizontalArrangement = Arrangement.SpaceBetween,
            verticalAlignment = Alignment.CenterVertically
        ) {
            Icon(
                Icons.Filled.LocalFireDepartment,
                contentDescription = null,
                tint = MaterialTheme.colors.secondary.copy(alpha = if (result.fastestLap?.rank == "1") 1f else 0f)
            )
            Column(
                modifier = Modifier.size(30.dp)
                    .clip(RoundedCornerShape(15.dp))
                    .padding(start = 2.dp, end = 10.dp, top = 2.dp, bottom = 2.dp),
                horizontalAlignment = Alignment.CenterHorizontally,
                verticalArrangement = Arrangement.Center
            ) {
                Text(result.points ?: "-", fontSize = 14.sp, fontWeight = FontWeight.Bold)
            }
        }
    }
}

File: swiftApp/Shared/Presentation/Common/DriverResultRow.swift
import SwiftUI
import shared

struct DriverResultRow: View {
    var result: ResultType
    var noDriverInfo: Bool = false
    
    var body: some View {
        HStack(alignment: .center) {
            HStack(alignment: .center) {
                StandingPosition(position: result.positionText)
                VStack(alignment: .center) {
                    HStack(alignment: .center) {
                        if !noDriverInfo {
                            Text("\(result.driver.givenName) \(result.driver.familyName)")
                                .font(.system(size: 18))
                                .fontWeight(.bold)
                                .frame(minWidth: 0, maxWidth: .infinity, alignment: .leading)
                                .padding(.horizontal, 10)
                                .padding(.vertical, 2)
                            VStack {
                                VStack(alignment: .center) {
                                    if let number = result.number {
                                        Text(number)
                                            .font(.system(size: 14))
                                            .fontWeight(.medium)
                                    }
                                }
                                .size(26)
                                .overlay(RoundedRectangle(cornerRadius: 13).stroke(lineWidth: 1))
                            }
                            .padding(.horizontal, 10)
                            .padding(.vertical, 2)
                        }
                    }
                    .fillMaxWidth()
                    HStack(alignment: .center) {
                        Text(result.constructor.name)
                            .font(.system(size: 16))
                            .fontWeight(.bold)
                            .frame(minWidth: 0, maxWidth: .infinity, alignment: .leading)
                            .padding(.horizontal, 10)
                            .padding(.vertical, 2)
                        VStack {
                            Text(result.time?.time ?? result.status)
                                .font(.system(size: 14))
                                .fontWeight(.medium)
                                .padding(.trailing, 10)
                        }
                        .padding(.vertical, 2)
                    }
                    .fillMaxWidth()
                }
                .fillMaxSize()
            }
            .fillMaxSize()
            HStack(alignment: .center) {
                #if os(iOS)
                if let uiImage = UIImage(named: "baseline_local_fire_department_black_24pt") {
                    Image(uiImage: uiImage)
                        .foregroundColor(Color.secondary.opacity(result.fastestLap?.rank == "1" ? 1 : 0))
                }
                #elseif os(macOS)
                if let nsImage = NSImage(named: "baseline_local_fire_department_black_24pt") {
                    Image(nsImage: nsImage)
                        .foregroundColor(Color.secondary.opacity(result.fastestLap?.rank == "1" ? 1 : 0))
                }
                #endif
                VStack(alignment: .center) {
                    Text(result.points ?? "-")
                        .font(.system(size: 14))
                        .fontWeight(.bold)
                }
                .size(30)
                .clipShape(Circle())
                .padding(.init(top: 2, leading: 2, bottom: 2, trailing: 10))
            }
        }
        .fillMaxWidth()
    }
}
